libraries/adaptive-expressions/src/triggerTrees/clause.ts

```diff
@@ -0,0 +1,285 @@
+export class Clause extends Expression {
+    private _ignored: Expression;
+
+    /**
+     * Initializes a new instance of the `Clause` class.
+     * @param clauseOrExpression A clause, expression or an array of expressions to initialize a `Clause`.
+     */
+    public constructor(clauseOrExpression?: Clause | Expression | Expression[]) {
+        super(ExpressionType.And, undefined);
+        if (clauseOrExpression) {
+            if (Array.isArray(clauseOrExpression)) {
+                const children: Expression[] = clauseOrExpression;
+                this.children = children;
+            } else if (clauseOrExpression instanceof Clause) {
+                const fromClause: Clause = clauseOrExpression;
+                this.children = [...fromClause.children];
+                for (const key in fromClause.anyBindings) {
+                    this.anyBindings[key] = fromClause.anyBindings[key];
+                }
+            } else if (clauseOrExpression instanceof Expression) {
+                const expression: Expression = clauseOrExpression;
+                this.children.push(expression);
+            }
+        }
+    }
+
+    /**
+     * Gets or sets the anyBinding dictionary.
+     */
+    public anyBindings: { [key: string]: string } = {};
+
+    /**
+     * Gets or sets whether the clause is subsumed.
+     */
+    public subsumed = false;
+
+    /**
+     * Gets a string that represents the current clause.
+     * @param builder An array of string to build the string of clause.
+     * @param indent An integer represents the number of spaces at the start of a line.
+     */
+    public toString(builder: string[] = [], indent = 0): string {
+        builder.push(' '.repeat(indent));
+        if (this.subsumed) {
+            builder.push('*');
+        }
+
+        builder.push('(');
+        let first = true;
+        for (const child of this.children) {
+            if (first) {
+                first = false;
+            } else {
+                builder.push(' && ');
+            }
+            builder.push(child.toString());
+        }
+
+        builder.push(')');
+        if (this._ignored) {
+            builder.push(' ignored(');
+            builder.push(this._ignored.toString());
+            builder.push(')');
+        }
+
+        for (const key in this.anyBindings) {
+            builder.push(` ${ key }->${ this.anyBindings[key] }`);
+        }
+        return builder.join('');
+    }
+
+    /**
+     * Compares the current `Clause` with another `Clause`.
+     * @param other The other `Clause` to compare.
+     * @param comparers A comparer, which is a dictionary of `PredicateComparer` with string keys.
+     * @returns A `RelationshipType` value between two `Clause` instances.
+     */
+    public relationship(other: Clause, comparers: { [name: string]: PredicateComparer }): RelationshipType {
+        let soFar: RelationshipType = RelationshipType.incomparable;
+        let shorter: Clause = this as Clause;
```

Comment:
I don't think you need both the `: Clause` and `as Clause` piece. Typescript will infer the type when using `as`.